tests/util/test_clem_array_functions.py

```diff
@@ -0,0 +1,466 @@
+    LUT,
+    convert_array_dtype,
+    convert_to_rgb,
+    estimate_int_dtype,
+    flatten_image,
+    get_dtype_info,
+    get_valid_dtypes,
+    merge_images,
+    shrink_value,
+    stretch_image_contrast,
+)
+
+known_dtypes = (
+    "int8",
+    "int16",
+    "int32",
+    "int64",
+    "uint8",
+    "uint16",
+    "uint32",
+    "uint64",
+    "float16",
+    "float32",
+    "float64",
+    "float128",
+    "complex64",
+    "complex128",
+    "complex256",
+)
+
+
+def test_get_valid_dtypes():
+    # All known dtypes should be generated by the function
+    assert len(set(known_dtypes) - set(get_valid_dtypes())) == 0
+
+
+@pytest.mark.parametrize("dtype", known_dtypes)
+def test_get_dtype_info(dtype: str):
+
+    # Load dtype info from NumPy
+    dtype_info = (
+        np.iinfo(dtype) if dtype.startswith(("int", "uint")) else np.finfo(dtype)
+    )
+    # Load dtype info via function
+    dtype_func = get_dtype_info(dtype)
+
+    # Compare the contents of the two objects
+    assert (
+        len(
+            {dtype_info.min, dtype_info.max, str(dtype_info.dtype)}
+            - {dtype_func.min, dtype_func.max, str(dtype_func.dtype)}
+        )
+        == 0
+    )
+
+
+dtype_estimation_test_matrix = (
+    # Min value | Max value | Bit depth (optional) | As float? (bool) | Expected estimate
+    # Test bit depth parameter
+    (-128, 127, 8, True, "int8"),
+    (0, 127, 16, False, "uint16"),
+    (0, 127, 32, True, "uint32"),
+    (-128, 127, 64, False, "int64"),
+    # Test auto-find ability
+    (0, 127, None, False, "uint8"),
+    (-32768, 32767, None, True, "int16"),
+    (-2147483648, 2147483647, None, False, "int32"),
+    (0, 9223372036854775807, None, True, "uint64"),
+)
+
+
+@pytest.mark.parametrize("test_params", dtype_estimation_test_matrix)
+def test_estimate_int_dtype(test_params: tuple[int, int, Optional[int], bool, str]):
+    vmin, vmax, bits, is_float, target = test_params
+
+    # Create test array
+    shape = (32, 32)
+    dtype = "float64" if is_float is True else "int64"
+    arr = np.random.randint(vmin, vmax, shape).astype(dtype)
+
+    estimate = estimate_int_dtype(arr, bits)
+    assert estimate == target
+
+
+@pytest.mark.parametrize("dtype", known_dtypes)
+def test_shrink_value(dtype: str):
+
+    # Encase in a "try-except" block to test expected fail conditions too
+    try:
+
+        # Get limits for the current array type
+        vmin = get_dtype_info(dtype).min
+        vmax = get_dtype_info(dtype).max
+
+        # Find new limits after shrinking
+        vmin_new = shrink_value(vmin)
+        vmax_new = shrink_value(vmax)
+        # Check that new values are within range allowed by array
+        assert float(vmin_new) >= vmin and float(vmax_new) <= vmax
+
+        # Check that they can be cast to the array properly
+        np.ones(1).astype(dtype) * vmax_new
+        np.ones(1).astype(dtype) * vmin_new
+        assert True
+
+    # Floats and complexes should fail
+    except TypeError:
```

Comment:
This probably doesn't quite test what you want. If an int also gave "TypeError" this test would still pass